lib/mongo/operation/shared/read_preference_supported.rb

```diff
@@ -36,47 +36,44 @@ module ReadPreferenceSupported
       #
       # @since 2.0.0
       def options(connection)
-        add_slave_ok_flag_maybe(super, connection)
+        options = super
+        if add_slave_ok_flag?(connection)
+          flags = options[:flags]&.dup || []
+          flags << :slave_ok
+          options = options.merge(flags: flags)
+        end
+        options
       end
 
-      # Adds :slave_ok flag to options based on the read preference specified
-      # in the operation or implied by the topology that the connection's
-      # server is a part of.
+      # Whether to add the :slave_ok flag to the request based on the
+      # read preference specified in the operation or implied by the topology
+      # that the connection's server is a part of.
       #
-      # @param [ Hash ] options The options calculated so far.
       # @param [ Server::Connection ] connection The connection that the
       #   operation will be executed on.
       #
-      # @return [ Hash ] The new options.
-      def add_slave_ok_flag_maybe(options, connection)
-        add_flag =
-          # https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#topology-type-single
-          if connection.description.standalone?
-            # Read preference is never sent to standalones.
-            false
-          elsif connection.server.cluster.single?
-            # In Single topology the driver forces primaryPreferred read
-            # preference mode (via the slave_ok flag, in case of old servers)
-            # so that the query is satisfied.
-            true
-          else
-            # In replica sets and sharded clusters, read preference is passed
-            # to the server if one is specified by the application, and there
-            # is no default.
-            read && read.slave_ok?
-          end
-
-        if add_flag
-          options= options.dup
-          (options[:flags] ||= []) << :slave_ok
+      # @return [ true | false ] Whether the :slave_ok flag should be added.
+      def add_slave_ok_flag?(connection)
+        # https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#topology-type-single
+        if connection.description.standalone?
+          # Read preference is never sent to standalones.
+          false
+        elsif connection.server.cluster.single?
+          # In Single topology the driver forces primaryPreferred read
+          # preference mode (via the slave_ok flag, in case of old servers)
+          # so that the query is satisfied.
+          true
+        else
+          # In replica sets and sharded clusters, read preference is passed
+          # to the server if one is specified by the application, and there
+          # is no default.
+          read && read.slave_ok? || false
```

Comment:
Question without knowing internal semantics, why does this require `|| false` to be included? Asking because `A || false` will always evaluate to `A` in boolean algebra, so I'm wondering about what Ruby does when `read` evaluates to a false-ish value.

Reply:
The `|| false` converts nil to false (when `read` is nil, in this case).